src/conn/conn_chunkcache.c

```diff
@@ -0,0 +1,346 @@
+            ret = -1;
+            goto err;
+        }
+
+        __wt_free(session, entry);
+        entry = NULL;
+    }
+
+err:
+    if (cursor != NULL)
+        WT_TRET(cursor->close(cursor));
+    if (entry != NULL)
+        __wt_free(session, entry);
+    return (ret);
+}
+
+/*
+ * __chunkcache_metadata_server --
+ *     Dispatch chunks of work (or stop the server) whenever we're signalled to do so.
+ */
+static WT_THREAD_RET
+__chunkcache_metadata_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    uint64_t cond_time_us;
+    bool signalled;
+
+    session = arg;
+    conn = S2C(session);
+    cond_time_us = 1 * WT_MILLION;
+
+    for (;;) {
+        /* Wait until the next event. */
+        __wt_cond_wait_signal(session, conn->chunkcache_metadata_cond, cond_time_us,
+          __chunkcache_metadata_run_chk, &signalled);
+
+        if (!__chunkcache_metadata_run_chk(session))
+            break;
+
+        if (signalled)
+            WT_ERR(__chunkcache_metadata_work(session));
+    }
+
+    if (0) {
+err:
+        WT_IGNORE_RET(__wt_panic(session, ret, "%s", "chunk cache metadata server error"));
+    }
+
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_chunkcache_metadata_create --
+ *     Start the server component of the chunk cache metadata subsystem.
+ */
+int
+__wt_chunkcache_metadata_create(WT_SESSION_IMPL *session)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    char *metadata_config;
+
+    conn = S2C(session);
+    chunkcache = &conn->chunkcache;
+
+    if (!F_ISSET(chunkcache, WT_CHUNKCACHE_CONFIGURED) || chunkcache->type != WT_CHUNKCACHE_FILE)
+        return (0);
+
+    /* Retrieve the chunk cache metadata config, and ensure it matches our startup config. */
+    ret = __chunkcache_get_metadata_config(session, &metadata_config);
+    if (ret == WT_NOTFOUND) {
+        WT_RET(__chunkcache_create_metadata_file(
+          session, chunkcache->capacity, chunkcache->hashtable_size, chunkcache->chunk_size));
+        __wt_verbose(session, WT_VERB_CHUNKCACHE, "%s", "created chunkcache metadata file");
+    } else if (ret == 0) {
+        WT_RET(__chunkcache_verify_metadata_config(session, metadata_config, chunkcache->capacity,
+          chunkcache->hashtable_size, chunkcache->chunk_size));
+        __wt_verbose(session, WT_VERB_CHUNKCACHE, "%s", "reused chunkcache metadata file");
+    } else {
+        WT_RET(ret);
+    }
+    __wt_free(session, metadata_config);
+
+    /* Start the internal thread. */
+    WT_ERR(__wt_cond_alloc(session, "chunkcache metadata", &conn->chunkcache_metadata_cond));
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_CHUNKCACHE_METADATA);
+
+    /* TODO set isolation */
+    WT_ERR(__wt_open_internal_session(
+      conn, "chunkcache-metadata-server", true, 0, 0, &conn->chunkcache_metadata_session));
+    session = conn->chunkcache_metadata_session;
```

Comment:
Shouldn't we do the following in the `__chunkcache_metadata_server` function?

```
    /* Start the internal thread. */
    WT_ERR(__wt_cond_alloc(session, "chunkcache metadata", &conn->chunkcache_metadata_cond));
    FLD_SET(conn->server_flags, WT_CONN_SERVER_CHUNKCACHE_METADATA);

    /* TODO set isolation */
    WT_ERR(__wt_open_internal_session(
      conn, "chunkcache-metadata-server", true, 0, 0, &conn->chunkcache_metadata_session));
    session = conn->chunkcache_metadata_session;
```

Reply:
Once we do the above we could get rid of the following - 
```
    if (0) {
err:
        FLD_CLR(conn->server_flags, WT_CONN_SERVER_CHUNKCACHE_METADATA);
        WT_TRET(__wt_chunkcache_metadata_destroy(session));
    }
```

Reply:
No, we can't move this code into that function - we need to pass the internal session to `__wt_thread_create`, and since `__chunkcache_metadata_server` runs in the context of that new thread, it's too late to create the new session there.